### ISSUE TITLE
Consider switching `synom` from `IResult<I, O>` to `Result<(I, O), ...>`

### ISSUE THREAD

**Issue by mystor** [dtolnay/syn#151]
Right now we use a custom result type for `synom`. In `nom` this was done to support the non-error `Incomplete` variant, which we don't use in `syn`. We don't have that need in synom, as we only have the successful and unsuccessful parse conditions. We should consider replacing our `enum IResult` with:

```rust
type Result<'a, O> = Result<(&'a [TokenStream], O), ParseError>
```

We could also make a new type instead of using the tuple on the Ok() side of the result. If we do that I would probably want to add `fn parse_ok<O>(rest: &[TokenStream], val: O) -> Result<O>` and `fn parse_err<O>() -> Result<O>` helper methods for ergonomic reasons.

For #148 I will probably change the input type from `&[TokenStream]` to a `Cursor` type.

**Comment by dtolnay** [comment]
I agree. This is long overdue.

**Comment by mystor** [comment]
I'll do this before #148, as it'll make that patch easier.

**Comment by dtolnay** [comment]
You are only working on this for syn/synom right? Proc-macro2 might benefit from this too https://github.com/alexcrichton/proc-macro2/issues/12.

**Comment by mystor** [comment]
Yes, I just started writing the patch for `syn/synom`. I wasn't intending to make the change to `proc-macro2`, but we should do it there as well.

**Comment by dtolnay** [comment]
Fixed in #152.
